hangups/channel.py

```diff
@@ -303,47 +303,33 @@ def _longpoll_request(self):
             res = yield from asyncio.wait_for(aiohttp.request(
                 'get', CHANNEL_URL_PREFIX.format('channel/bind'),
                 params=params, cookies=self._cookies, headers=headers,
-                connector=self._connector
-            ), CONNECT_TIMEOUT)
+                connector=self._connector), CONNECT_TIMEOUT)
+
+            if res.status != 200:
+                if res.status == 400 and res.reason == 'Unknown SID':
+                    raise UnknownSIDError('SID became invalid')
+                raise exceptions.NetworkError(
+                    'Request return unexpected status: {}: {}'.format(
+                        res.status, res.reason))
+
+            while True:
+                chunk = yield from asyncio.wait_for(
+                    res.content.read(MAX_READ_BYTES), PUSH_TIMEOUT)
+                if not chunk:
+                    break
+
+                yield from self._on_push_data(chunk)
+
         except asyncio.TimeoutError:
             raise exceptions.NetworkError('Request timed out')
-        except aiohttp.ServerDisconnectedError as e:
-            raise exceptions.NetworkError('Server disconnected error: {}'
-                                          .format(e))
-        except aiohttp.ClientError as e:
-            raise exceptions.NetworkError('Request connection error: {}'
-                                          .format(e))
-        if res.status == 400 and res.reason == 'Unknown SID':
-            raise UnknownSIDError('SID became invalid')
-        elif res.status != 200:
+        except aiohttp.ServerDisconnectedError as err:
             raise exceptions.NetworkError(
-                'Request return unexpected status: {}: {}'
-                .format(res.status, res.reason)
-            )
-        while True:
-            try:
-                chunk = yield from asyncio.wait_for(
-                    res.content.read(MAX_READ_BYTES), PUSH_TIMEOUT
-                )
-            except asyncio.TimeoutError:
-                raise exceptions.NetworkError('Request timed out')
-            except aiohttp.ServerDisconnectedError as e:
-                raise exceptions.NetworkError('Server disconnected error: {}'
-                                              .format(e))
-            except aiohttp.ClientError as e:
-                raise exceptions.NetworkError('Request connection error: {}'
-                                              .format(e))
-            except asyncio.CancelledError:
-                # Prevent ResourceWarning when channel is disconnected.
+                'Server disconnected error: %s' % err)
+        except aiohttp.ClientError as err:
+            raise exceptions.NetworkError('Request connection error: %s' % err)
+        finally:
+            if 'res' in locals():
```

Comment:
https://github.com/tdryer/hangups/pull/326#discussion_r124706465 still leaves the response in this exit open https://gist.github.com/das7pad/e8f63175377576a0cb3466d20a417fa5#file-main-py-L25..L26:
a ``asyncio.CancelledError`` between creating the response and reading from it.

We could set ``response = None`` before creating the request and check it in a finally block. This would drop the usage of ``locals()`` and cover all cases.
```
response = None
try:
    response = yield from aiohttp.request('GET', TEST_URL)
    text = yield from response.text()
except aiohttp.ClientError:
    raise ValueError('some hangups error')
finally:
    if response is not None:
        response.release()
```